cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.api.metadata;
+
+import java.util.Map;
+import java.util.Set;
+import javax.ws.rs.NotFoundException;
+
+/**
+ * The context for reading metadata from program
+ */
+public interface MetadataReaderContext {
+
+  /**
+   * Returns a set of {@link MetadataRecord} representing all metadata (including properties and tags) for the specified
+   * {@link MetadataEntity} in both {@link MetadataScope#USER} and {@link MetadataScope#SYSTEM}.
+   *
+   * @throws NotFoundException if the specified entity was not found
+   */
+  Set<MetadataRecord> getMetadata(MetadataEntity metadataEntity) throws NotFoundException;
+
+  /**
+   * Returns a set of {@link MetadataRecord} representing all metadata (including properties and tags) for the specified
+   * {@link MetadataEntity} in the specified {@link MetadataScope}.
+   *
+   * @throws NotFoundException if the specified entity was not found
+   */
+  Set<MetadataRecord> getMetadata(MetadataScope scope, MetadataEntity metadataEntity) throws NotFoundException;
+
+  /**
+   * @return all the properties for the specified {@link MetadataEntity} in both {@link MetadataScope#USER} and
+   * {@link MetadataScope#SYSTEM}
+   * @throws NotFoundException if the specified entity was not found
+   */
+  Map<MetadataScope, Map<String, String>> getProperties(MetadataEntity metadataEntity) throws NotFoundException;
+
+  /**
+   * @return a {@link Map} representing the metadata of the specified {@link MetadataEntity} in the specified
+   * {@link MetadataScope}
+   * @throws NotFoundException if the specified entity was not found
+   */
+  Map<String, String> getProperties(MetadataScope scope, MetadataEntity metadataEntity) throws NotFoundException;
+
+  /**
+   * @return all the tags for the specified {@link MetadataEntity} in both {@link MetadataScope#USER} and
+   * {@link MetadataScope#SYSTEM}
+   * @throws NotFoundException if the specified entity was not found
+   */
+  Map<MetadataScope, Set<String>> getTags(MetadataEntity metadataEntity) throws NotFoundException;
+
+  /**
+   * @return all the tags for the specified {@link MetadataEntity} in the specified {@link MetadataScope}
+   * @throws NotFoundException if the specified entity was not found
+   */
+  Set<String> getTags(MetadataScope scope, MetadataEntity metadataEntity) throws NotFoundException;
+}
```

Comment:
Is it necessary to have a separate method for tags and properties? Is there a performance impact? Or does it harm to just always return all metadata records?

Reply:
We have separate methods in the our clients and rest apis so the ideas is taken from there for consistency of apis. Although, yes so many apis just to expose different part of the same data doesnt seem right. There is no performance impact as such because all these apis read the hbase table to get the metadata (including properties and tags) internally and then filter out whats needed as post processing. So we can just expose one which gives the MetadataRecord.

But now the idea is that we will announce the MetadataEntity whose metadata we want to be available in the context of initialize. If we do the above then we will only support making the whole MetadataRecord available in the context even if the user wants just the properties of tags. This might lead to filling up more memory than required.